File: core/src/main/java/ru/oseval/datahub/java/MemoryStorage.java
package ru.oseval.datahub.java;

import java.util.Comparator;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Function;

public class MemoryStorage implements Datahub.Storage {
    private ConcurrentMap<String, Object> ids = new ConcurrentHashMap<String, Object>();

    @Override
    public void register(String entityId, Object dataClock, Function<Void, Void> callback) {
        ids.put(entityId, dataClock);
        callback.apply(null);
    }

    @Override
    public <C> void increase(String entityId, C dataClock, Comparator<C> cmp, Function<Void, Void> callback) {
        Object res = ids.computeIfPresent(entityId, (k, _clk) -> {
            C clk = (C) _clk;
            return cmp.compare(dataClock, clk) > 0 ? dataClock : clk;
        });

        if (res == null) {
            ids.putIfAbsent(entityId, dataClock);
        }
    }

    @Override
    public void getLastClock(String entityId, Function<Optional<Object>, Void> callback) {
        callback.apply(Optional.ofNullable(ids.get(entityId)));
    }
}

File: core/src/main/scala/ru/oseval/datahub/MemoryStorage.scala
package ru.oseval.datahub

import java.util.concurrent.ConcurrentHashMap

import ru.oseval.datahub.AsyncDatahub.Storage

import scala.concurrent.Future

class MemoryStorage extends Storage {
  private val ids: ConcurrentHashMap[String, Any] = new ConcurrentHashMap[String, Any]

  override def increase(entity: Entity)(dataClock: entity.ops.D#C): Future[Unit] = {
    val res = ids.computeIfPresent(entity.id, { case (k: String, _clk: Any) =>
      entity.ops.matchClock(_clk).filter(entity.ops.ordering.gt(_, dataClock)) getOrElse dataClock
    })

    if (res == null) ids.putIfAbsent(entity.id, dataClock)

    Future.unit
  }

  override def getLastClock(entity: Entity): Future[Option[entity.ops.D#C]] =
    Future.successful(Option(ids.get(entity.id)).flatMap(entity.ops.matchClock(_)))
}
